### ISSUE TITLE
Compatibility With Arm7

### ISSUE THREAD

**Issue by NoahZuo** [nfrechette/acl-ue4-plugin#90]
This error showed up when building the project with arm-7: 

![image](https://user-images.githubusercontent.com/10735147/124431202-a8b7f280-dda2-11eb-8595-77fe940feef6.png)

It seems that something is wrong with rtm.

**Comment by nfrechette** [comment]
`std::copysign` is part of C++11, see [here](https://en.cppreference.com/w/cpp/numeric/math/copysign).
It comes with the `cmath` header and is included in that file [here](https://github.com/nfrechette/rtm/blob/dda7c7452a37c859d1805a15d104e01173df6d08/includes/rtm/scalard.h#L34).

All versions of RTM and ACL are compiled for ArmV7 and the unit test/regression tests are ran manually as well on Android with and without SIMD support. However, I do not always test within UE4 since Armv7 is pretty old now and no longer supported by the Google Play store.

By the looks of it, the version of UE4 that you are using must have an old android toolchain that does not properly support C++11. Since C++11 is over 10 years old at this point, I'm afraid you'll have to make custom changes to RTM to fix this error on your end. You could try what the compiler suggests and remove the `std::` prefix to call the C version.

Another to fix you could try is to polyfill. Can you try the following and let me know if it works at the top of scalard.h after the cmath include?

```c++
#if defined(__GNUG__) && !defined(_LIBCPP_VERSION) && !defined(_GLIBCXX_USE_CXX11_ABI)
namespace std
{
    using ::copysign;
}
#endif
```

This checks that we are using GCC when compiling C++, we check that we aren't using `libc++` which is the LLVM C++ standard library (clang defines a lot of GCC macros), and we check the USE CXX11 macro which is defined only starting with GCC 5. I suspect you are using a GCC version that is 4.x where C++11 support was partial. For example, see [here](https://stackoverflow.com/questions/22922961/c11-cmath-functions-not-in-std-namespace-for-android-ndk-w-gcc-4-8-or-clang-3). This is an issue I previously ran into with sjson-cpp and I used the macros above to polyfill. If this works, I will add it to RTM but a fix will not be rolled out anytime soon as part of the UE4 plugin. GCC 5.1 came out in April 2015 and supporting anything older than that isn't a very high priority at this time.

**Comment by NoahZuo** [comment]
> 
> 
> `std::copysign` is part of C++11, see [here](https://en.cppreference.com/w/cpp/numeric/math/copysign).
> It comes with the `cmath` header and is included in that file [here](https://github.com/nfrechette/rtm/blob/dda7c7452a37c859d1805a15d104e01173df6d08/includes/rtm/scalard.h#L34).
> 
> All versions of RTM and ACL are compiled for ArmV7 and the unit test/regression tests are ran manually as well on Android with and without SIMD support. However, I do not always test within UE4 since Armv7 is pretty old now and no longer supported by the Google Play store.
> 
> By the looks of it, the version of UE4 that you are using must have an old android toolchain that does not properly support C++11. Since C++11 is over 10 years old at this point, I'm afraid you'll have to make custom changes to RTM to fix this error on your end. You could try what the compiler suggests and remove the `std::` prefix to call the C version.
> 
> Another to fix you could try is to polyfill. Can you try the following and let me know if it works at the top of scalard.h after the cmath include?
> 
> ```c++
> #if defined(__GNUG__) && !defined(_LIBCPP_VERSION) && !defined(_GLIBCXX_USE_CXX11_ABI)
> namespace std
> {
>     using ::copysign;
> }
> #endif
> ```
> 
> This checks that we are using GCC when compiling C++, we check that we aren't using `libc++` which is the LLVM C++ standard library (clang defines a lot of GCC macros), and we check the USE CXX11 macro which is defined only starting with GCC 5. I suspect you are using a GCC version that is 4.x where C++11 support was partial. For example, see [here](https://stackoverflow.com/questions/22922961/c11-cmath-functions-not-in-std-namespace-for-android-ndk-w-gcc-4-8-or-clang-3). This is an issue I previously ran into with sjson-cpp and I used the macros above to polyfill. If this works, I will add it to RTM but a fix will not be rolled out anytime soon as part of the UE4 plugin. GCC 5.1 came out in April 2015 and supporting anything older than that isn't a very high priority at this time.

Sadly this does not work. I tried to fix this issue by myself yesterday but it turned out that there are still problems like this: 
![image](https://user-images.githubusercontent.com/10735147/124536126-fcc9e200-de49-11eb-82a9-d3ba55e687e7.png)

I've tried your polyfill fix, it is still the same issue.

**Comment by nfrechette** [comment]
It looks like the polyfill only worked partially. It worked for `double` but not for `float` since C math doesn't define that override.

Try to polyfill like this instead:

```c++
#if defined(__GNUG__) && !defined(_LIBCPP_VERSION) && !defined(_GLIBCXX_USE_CXX11_ABI)
namespace std
{
    inline float copysign(float x, float y) { return ::copysignf(x, y); }
    inline double copysign(double x, double y) { return ::copysign(x, y); }
}
#endif
```

That should setup the proper overrides.
For the uninitialized `result`, you can initialize it like this: `result = lhs;`
The result value is entirely overwritten so it can be initialized with anything to silence the warning with 0 overhead.

I only found GCC 4.8 on Ubuntu 18 and although there are some warnings in compression, those are stripped for android in UE4. I couldn't reproduce the above issues even with no SIMD on x64. I presume I would need to setup your particular NDK version but that might not be so easy. I'll see what I can do with my old branch of UE 4.24, hopefully it uses the same NDK and toolchain.

**Comment by nfrechette** [comment]
I was able to reproduce the copysign issue in UE 4.24 by hacking the plugin in but the uninitialized variable warning wasn't present.
I made a fix for both issues [here](https://github.com/nfrechette/rtm/pull/114) in RTM. It will end up in the plugin at some point but not anytime soon. In the meantime, please integrate the fix manually.

If the fix works for you, we'll close this issue as resolved.

**Comment by nfrechette** [comment]
I pushed another commit to fix the GCC 4.9 support issues, see https://github.com/nfrechette/rtm/pull/114

**Comment by NoahZuo** [comment]
As an alternation, I've post my result to that issue. Please check that out.
